### PR TITLE
[feature-wip](array-type) explode support more sub types

### LINKED ISSUE #10052: [Feature] Add lots of Array Functions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

We have supported `array` data type recently, all pr(s) about `array` please refer to https://github.com/apache/incubator-doris/issues/7570 .

Now we need to add lots of array functions to make `array` more powerful.
If anybody is interesting to `array`, please let me know.

Array Functions DONE:
- size/cardinality
- array_position
- array_contains
- element_at

Array Functions DOING:
- array_sum
- array_avg
- array_min
- array_max
- array_product
- array_distinct
- array_remove
- arrays_overlap

Array Functions TODO:
- array_union
- array_intersect
- array_except
- reverse
- concat_ws
- array_join
- array_sort
- array_slice

### Use case

_No response_

### Related issues

https://github.com/apache/incubator-doris/issues/7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by cambyzju**
# Proposed changes

Issue Number: close #10052

## Problem Summary:

refractor explode and explode_outer:
1. support more sub types;
2. support nullable elements of array;

Usage example:
```
mysql> set enable_vectorized_engine = true
mysql> set enable_array_type = true

mysql> select e1 from (select 1 k1) as t lateral view explode([1,2,3]) tmp1 as e1;
+------+
| e1   |
+------+
|    1 |
|    2 |
|    3 |
+------+

mysql> select e1 from (select 1 k1) as t lateral view explode_outer(null) tmp1 as e1;
+------+
| e1   |
+------+
| NULL |
+------+

mysql> select e1 from (select 1 k1) as t lateral view explode([]) tmp1 as e1;
Empty set (0.010 sec)

mysql> select e1 from (select 1 k1) as t lateral view explode([null,1,null]) tmp1 as e1;
+------+
| e1   |
+------+
| NULL |
|    1 |
| NULL |
+------+

mysql> select e1 from (select 1 k1) as t lateral view explode_outer([null,1,null]) tmp1 as e1;
+------+
| e1   |
+------+
| NULL |
|    1 |
| NULL |
+------+
```

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yangzhg** [comment]
maybe need rebase to fix p0

File: be/src/vec/exprs/table_function/vexplode.cpp [hunk 1]
```diff
@@ -56,26 +48,22 @@ Status VExplodeTableFunction::process_row(size_t row_idx) {
     DCHECK(row_idx < _array_column->size());
     _is_current_empty = false;
     _eos = false;
+    _cur_offset = 0;
+    _array_off = (*_detail.offsets_ptr)[row_idx - 1];
+    _cur_size = (*_detail.offsets_ptr)[row_idx] - _array_off;
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/exprs/table_function/vexplode.cpp]: table function framework also need `_cur_size`, so we set it here.

File: be/src/vec/exprs/table_function/vexplode.h [hunk 2]
```diff
@@ -39,9 +40,9 @@ class VExplodeTableFunction : public TableFunction {
     virtual Status get_value_length(int64_t* length) override;
 
 private:
-    const UInt8* _array_null_map;
-    const ColumnArray* _array_column;
-    size_t _pos;
+    ColumnPtr _array_column;
+    ColumnArrayExecutionData _detail;
+    size_t _array_off; // start offset of array[row_idx]
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply be/src/vec/exprs/table_function/vexplode.h]: done

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 3]
```diff
@@ -2711,11 +2711,10 @@ private void initTableFunction() {
                 "_ZN5doris19DummyTableFunctions22explode_numbersEPN9doris_udf15FunctionContextERKNS1_9IntValE");
 
         initTableFunctionListWithCombinator(EXPLODE);
-        addTableFunctionWithCombinator(EXPLODE, Type.INT, Function.NullableMode.ALWAYS_NULLABLE,
-                Lists.newArrayList(new ArrayType(Type.INT)), false,
-                "_ZN5doris19DummyTableFunctions7explodeEPN9doris_udf15FunctionContextERKNS1_13CollectionValE");
-        addTableFunctionWithCombinator(EXPLODE, Type.VARCHAR, Function.NullableMode.ALWAYS_NULLABLE,
-                Lists.newArrayList(new ArrayType(Type.VARCHAR)), false,
-                "_ZN5doris19DummyTableFunctions7explodeEPN9doris_udf15FunctionContextERKNS1_13CollectionValE");
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:2719]: What is the dummy table function?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java [hunk 4]
```diff
@@ -156,6 +157,23 @@ public abstract class Type {
...
+        arraySubTypes.add(BIGINT);
+        arraySubTypes.add(LARGEINT);
+        arraySubTypes.add(FLOAT);
+        arraySubTypes.add(DOUBLE);
+        arraySubTypes.add(VARCHAR);
+        arraySubTypes.add(CHAR);
+        arraySubTypes.add(DATE);
+        arraySubTypes.add(DATETIME);
+        arraySubTypes.add(DECIMALV2);
+        arraySubTypes.add(TIME);
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java:175]: TIME?

Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Type.java]: removed

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java (hunk 3 above)
Comment (cambyzju) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:2719]: after non-vec codes removed, dummy functions also can removed later.

**Review by xy720 — APPROVED** [review]
LGTM

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**